lib/Dialect/Torch/Transforms/ScalarizeShapes.cpp

```diff
@@ -898,33 +981,55 @@ class ScalarizeShapesPass : public ScalarizeShapesBase<ScalarizeShapesPass> {
-                    FoldAtenSqueezePattern, FoldAtenUnsqueezePattern,
-                    FoldAtenWhereSelf, CanonicalizeAtenViewPattern,
-                    PropagateAtenEqTensorPattern, PropagateAtenWhereSelfPattern,
-                    FoldAtenSqueezeDimPattern,
-                    RemoveUnusedPattern<Torch::AtenIntBoolOp>,
-                    RemoveUnusedPattern<Torch::AtenEqIntOp>,
-                    RemoveUnusedPattern<Torch::PrimNumToTensorScalarOp>,
-                    RemoveUnusedPattern<Torch::AtenFullOp>,
-                    RemoveUnusedPattern<Torch::AtenUnsqueezeOp>,
-                    RemoveUnusedPattern<Torch::AtenSqueezeDimOp>,
-                    RemoveUnusedPattern<Torch::AtenSizeIntOp>,
-                    RemoveUnusedPattern<Torch::AtenSliceTensorOp>,
-                    RemoveUnusedPattern<Torch::AtenTensorOp>,
-                    RemoveUnusedPattern<Torch::ConstantBoolOp>,
-                    RemoveUnusedPattern<Torch::ConstantIntOp>,
-                    RemoveUnusedPattern<Torch::ConstantNoneOp>,
-                    RemoveUnusedPattern<Torch::PrimListConstructOp>>(context);
 
+    // populate patterns
+    populateScalarizationPropagationPatterns(patterns);
+    populateScalarizationFoldPatterns(patterns);
+    populateScalarizationCanonicalizePatterns(patterns);
+    populateScalarizationRemovePatterns(patterns);
     context->getLoadedDialect<mlir::arith::ArithDialect>()
         ->getCanonicalizationPatterns(patterns);
-    if (failed(applyPatternsAndFoldGreedily(getOperation(),
-                                            std::move(patterns)))) {
+
+    // walk func op to get a subgraph of shape computations
+    auto funcOp = getOperation();
+    llvm::SetVector<Operation *> shapeCalculationOps;
+    funcOp.walk<WalkOrder::PostOrder, mlir::ReverseIterator>(
+        [&](Operation *op) {
+          // walking the func op in reverse order, start adding ops when we
+          // reach an anchor point (a prim list of ints)
+          if (isPrimListOfInts(op)) {
+            shapeCalculationOps.insert(op);
+            return;
+          }
+          // add view ops for now until the decompositions for flatten and
+          // unflatten are removed.
+          if (isa<AtenViewOp>(op)) {
+            shapeCalculationOps.insert(op);
+            return;
+          }
+          // all ops which feed into the list-like anchor ops get added to the
+          // set. Stop when we feed into a source op for shape scalarization
+          for (OpOperand &use : op->getUses()) {
+            Operation *userOp = use.getOwner();
+            if (shapeCalculationOps.contains(userOp) &&
+                !isSourceOpForShapeScalarization(userOp) &&
+                !isa<AtenViewOp>(userOp)) {
+              shapeCalculationOps.insert(op);
+              return;
+            }
+          }
+        });
+
+    GreedyRewriteConfig config;
+    config.strictMode = GreedyRewriteStrictness::ExistingAndNewOps;
```

Comment:
Ok, so this is actually the key thing to this algorithm and needs some explanation: You are relying on patterns that, given some anchor, always rewrite to produce some new ops that (if there are further ops above) introduce an op that is a new anchor that goes on the worklist. So this is still a bottom up rewrite where each step is producing a locally correct graph transformation (at the cost of each step needing to create a number of new ops that will just be thrown away on the next pattern match).

This is a bit different from what I had described: do a first traversal to identify all dominators that end with an anchor. Then transform top-down, where each pattern produces an invalid graph transformation that does not need to materialize throw away ops at each step. In this approach, you would make one pass top down through the IR doing transformations and noting errors for anything that could not be converted.

Given the code that already exists for this and that the top-down way is definitely harder to write, doing it this way is fine. The cost will be more garbage/allocations that scale with the number of changes you have to make. And it is harder to do precise error handling. The up-side, though, is that if a pattern fails to fully scalarize, the graph is in a consistent but suboptimal state and can still be used.

So to answer the two questions you raised in the PR description:

1. PrimListOfInts may be an ok anchor. In my mind, that was an anchor but only if it was consumed as an index-oriented operand of an op. It may be a distinction without a difference in the end. I'm just paranoid because I know that in these kinds of graphs, there are definitely weird bottlenecks of such things that need more discrimination.
2. For error handling, if you don't want to do another walk, create your "Propagate*" pattern classes to take a reference to some PropagationState&. For things your patterns encounter that indicate a failure to convert that you want to report, add then to some set in there. Then at the end of your pass, emit one warning with multiple notes indicating the ops that could not be converted. The reason you want your patterns themselves to be silent is that patterns can be run multiple times during iteration and it is impossible to have precise control of your diagnostic behavior as a result.

Reply:
Thanks for the comments and offline discussion yesterday, Stella.

I added some more descriptive comments about the walk to collect the `shapeCalculationOps`, so hopefully this resolves some concerns mentioned in your first paragraph. I'd like to keep the graph transformations legal for now, since this will take some time to nail down all of the patterns we need to implement. Reading the debug printouts for the pass as it is, I think we traverse the IR and apply patterns in a safe and reasonably efficient way without having to invent a bespoke PatternRewriteDriver. 

1. I'll scan through the op registry for prim list of int inputs for ops and look for problematic examples. Should be able to remove those cases manually if they aren't common.
2. Sounds like a good plan for error handling.

Reply:
I updated the original comment with where I am at with this, so I'm going to resolve this conversation for now.